### ISSUE TITLE
Tips for artifact detection?

### ISSUE THREAD

**Issue by AndrewYatzkan** [neonbjb/tortoise-tts#91]
I've noticed that it's not uncommon for TorToiSe to generate weird artifacts in place of words or entire sentences.

I would like to use TorToiSe in a project where it's not possible to manually review every output, though, so does anybody have any suggestions for how to detect these artifacts so that I can re-generate the audio or revert to standard TTS?

**Comment by wavymulder** [comment]
I think you could feed the Tortoise output into Ocotillo (check nenobjb's github, it was used to generate the scripts for Tortoise's dataset), then compare that to your script. I'm not good at programming but I imagine it's possible.

**Comment by AndrewYatzkan** [comment]
> I think you could feed the Tortoise output into Ocotillo (check neonbjb's github, it was used to generate the scripts for Tortoise's dataset), then compare that to your script. I'm not good at programming but I imagine it's possible.

Thanks for the suggestion, however I'm hesitant to implement this because the transcription may not be an exact match if someone, say, has an accent. I suppose I could instead check that the prompt and transcription are a "close enough" match but I feel like there has to be a more direct solution.

**Comment by PiotrDabkowski** [comment]
If you had access to dvae encoder (converts audio to 8192 codes @22hz, that are being generated by the transformer), then you could train a simple ASR model on top of these discrete codes via CTC loss. And then compute the loss on the generated sequence (mismatch of generated codes vs requested text), and rejecting sequences where loss is too high.

Generating mulitple sequences and choosing the best one works OK, but is slow and not very reliable - what if you are unlucky and all are bad, and the chance for a good one is low? For more robust and faster solution, the logit choice could be guided during the code sampling.

**Comment by neonbjb** [comment]
Agreed that this would be a good approach. I'm very sorry that I can't release the DVAE so that others could play with something like this.

**Comment by PiotrDabkowski** [comment]
@neonbjb  Thats a shame, it was for some time on huggingface and I was able to reverse-engineer it, thought you simply considered it useless for others as it as it is not needed for TTS. But if it is your decision to not share then will not distribute. Your quantizer was super useful for my initial experiments and also for audio CE loss comparison of my model vs tortoise. Now moved to my own quantizer.

**Comment by neonbjb** [comment]
It's a critical component needed for fine-tuning and Tortoise is a deep-fake machine when fine-tuned. That's why I am with-holding it. Appreciate you doing the same.

**Comment by KMiNT21** [comment]
> weird artifacts in place of words or entire sentences

I use Wisper for checking output of any TTS.  :) Just compare results with input text. If difference more then X, check it manually.
